### PR TITLE
feat(loaderapi): create initial loader api

### PR DESCRIPTION
**Pull request by TheLarkInn**

### CODE REVIEW

File: content/api/loaders.md [hunk 1]
```diff
@@ -3,4 +3,302 @@ title: Loader API
...
+		// fast exit
+		return "module.exports = require(" + JSON.stringify("-!" + remainingRequest) + ");";
+	}
+	data.value = 42;
+};
+```
+
+## The loader context 
+
+The loader context represents the properties that are available inside of a loader assigned to the `this` property
```
Comment (SpaceK33z) [inline content/api/loaders.md]: End with a dot ;).

File: content/api/loaders.md [hunk 2]
```diff
@@ -3,4 +3,302 @@ title: Loader API
...
+and provide a powerful way to handle frontend build steps. Loaders can transform files from a different language like, CoffeeScript to JavaScript, or inline images as data URLs. Loaders even allow you to do things like require() css files right in your JavaScript!
+
+To tell webpack to transform a module with a loader, you can specify the loader in the webpack [configuration](../configuration) file (preferred) or in the module request, such as in a `require()` call.
+
+## How to write a loader
+
+A loader is just a file that exports a function. The compiler calls this function and passes the result of the previous loader or the resource file into it. The this context of the function is filled-in by the compiler with some useful methods that allow the loader to, among other things, change its invocation style to async or get query parameters. The first loader is passed one argument: the content of the resource file. The compiler expects a result from the last loader. The result should be a String or a Buffer (which is converted to a string), representing the JavaScript source code of the module. An optional SourceMap result (as JSON object) may also be passed.
```
Comment (SpaceK33z) [inline content/api/loaders.md]: s/The this/The `this`/
(I didn't know at first if you were talking about the JS `this` or real-life this 😄)

Comment (SpaceK33z) [inline content/api/loaders.md]: s/String/\`String\`/
s/Buffer/\`Buffer\`/

File: content/api/loaders.md [hunk 3]
```diff
@@ -3,4 +3,302 @@ title: Loader API
 sort: 4
 ---
 
-> TODO
+Loaders allow you to preprocess files as you require() or “load” them. Loaders are kind of like “tasks” in other build tools, 
+and provide a powerful way to handle frontend build steps. Loaders can transform files from a different language like, CoffeeScript to JavaScript, or inline images as data URLs. Loaders even allow you to do things like require() css files right in your JavaScript!
+
+To tell webpack to transform a module with a loader, you can specify the loader in the webpack [configuration](../configuration) file (preferred) or in the module request, such as in a `require()` call.
```
Comment (SpaceK33z) [inline content/api/loaders.md]: Is there another page that explains how to use a loader in configuration and in a `require()` call? If so, I'd suggest to link to that.

File: content/api/loaders.md [hunk 4]
```diff
@@ -3,4 +3,302 @@ title: Loader API
...
+
+### `target`
+
+Target of compilation. Passed from configuration options.
+
+Example values: `"web"`, `"node"`
+
+### `webpack`
+
+Set to true when this is compiled by webpack.
```
Comment (SpaceK33z) [inline content/api/loaders.md]: (saying as a user) what does this mean? aren't all loaders compiled by webpack?

**Comment by SpaceK33z** [comment]
Nice work @TheLarkInn!!

File: content/api/loaders.md (hunk 4 above)
Comment (TheLarkInn) [inline content/api/loaders.md]: Loader runner is a separate module so could be used externally. I should mark that.

File: content/api/loaders.md (hunk 1 above)
Comment (TheLarkInn) [inline content/api/loaders.md]: Done

File: content/api/loaders.md (hunk 3 above)
Comment (TheLarkInn) [inline content/api/loaders.md]: I don't believe those PR's have landed yet. I'll create an issue (after this merges). I'll also add a TODO. Done

File: content/api/loaders.md [hunk 5]
```diff
@@ -3,4 +3,302 @@ title: Loader API
 sort: 4
 ---
 
-> TODO
+Loaders allow you to preprocess files as you require() or “load” them. Loaders are kind of like “tasks” in other build tools,
```
Comment (TheLarkInn) [inline content/api/loaders.md]: done

File: content/api/loaders.md [hunk 6]
```diff
@@ -1,6 +1,310 @@
...
+
+By default, the resource file is treated as utf-8 string and passed as String to the loader. By setting raw to true the loader is passed the raw `Buffer`. Every loader is allowed to deliver its result as `String` or as `Buffer`. The compiler converts them between loaders.
+
+**raw-loader.js**
+
+```javascript
+module.exports = function(content) {
+	assert(content instanceof Buffer);
+	return someSyncOperation(content);
+	// return val`ue can` be a `Buffer` too
```
Comment (SpaceK33z) [inline content/api/loaders.md]: Small error with the backticks here, sorry for nitpicking haha.

**Comment by TheLarkInn** [comment]
Oh what the crawling assets about?

**Comment by SpaceK33z** [comment]
I *think* it has to do with an invalid link. Unfortunately it shows no information to narrow it down at all...

File: content/api/loaders.md (hunk 6 above)
Comment (TheLarkInn) [inline content/api/loaders.md]: No problem it's a good catch!!

**Comment by TheLarkInn** [comment]
@bebraw do you know what these errors might pertain to? @SpaceK33z and I both had passing builds locally for lint:links

**Comment by bebraw** [comment]
No idea. You would have to ask the hyperlink guys to get a better idea of the error. It could be something Node version related etc.

**Comment by TheLarkInn** [comment]
No worries, Im going to revert the change I just made as it looked like that gave more verbose error.

**Comment by TheLarkInn** [comment]
@SpaceK33z @bebraw looks like it was the /api/plugins link. Fixed!!!!!! Going to merge!

**Comment by TheLarkInn** [comment]
Fixes #21

**Pull request merged**